### DISCUSSION TITLE
Implementing TDD Resource Grid with AWGN channel module

### DISCUSSION THREAD

**Discussion by neilb22** [NVlabs/sionna#73] [Q&A] [answered]
Hi everyone.

I'm trying to generate a TDD resource grid and then pass it as an argument in the Mapper and Channel modules.

When using the OFDMChannel module, I can pass the resource grid as a parameter. But, it does not seem to be the case for the AWGN (or the Flat Fading) module(s). I have also noticed that the ofdm.ResourceGridMapper used in the examples returns a tensor with a shape different from that of the mapping.Mapper module that is used for the AWGN channel. 

If I may, I would **request some guidance on the approach to be taken so that I can incorporate the TDD grid that I have created into my AWGN model.**

To generate the TDD Resource Grid, I've used a modified version of the ofdm.ResourceGrid module, with the ` build_type_grid() ` method modified as follows:

```
        mask1 = 2*tf.ones(shape + [self._fft_size], tf.int32).numpy()
        for i in range(0, self._fft_size):
            if i%2==0:
                mask1[0,0,:,i] = 0
                
        mask = tf.convert_to_tensor(mask1)
```

The resource grid that is generated looks like this: 
![image](https://user-images.githubusercontent.com/114391942/206191164-9b4a9c67-936a-4ebb-b4af-0ef1f2f63195.png)

I am not able to figure out how to use this grid for the AWGN channel. Any feedback would be highly appreciated.

Kind Regards,
Neil

**Comment by jhoydis** [comment, marked as answer]
Hi Neil,

The [AWGN](https://nvlabs.github.io/sionna/api/channel.wireless.html#sionna.channel.AWGN) channel layer does not require knowledge of the [ResourceGrid](https://nvlabs.github.io/sionna/api/ofdm.html#sionna.ofdm.ResourceGrid) as it will simply add complex Gaussian noise to the elements of the input tensor (which can have arbitrary shape).

If you want to simulate a flat fading channel, you can use the [RayleighBlockFading](https://nvlabs.github.io/sionna/api/channel.wireless.html#sionna.channel.RayleighBlockFading) channel model together with the [OFDMChannel](https://nvlabs.github.io/sionna/api/channel.wireless.html#sionna.channel.OFDMChannel) layer.

Note that the [Mapper](https://nvlabs.github.io/sionna/api/mapping.html#sionna.mapping.Mapper) maps a sequence of bits to constellation symbols, while the [ResourceGridMapper](https://nvlabs.github.io/sionna/api/ofdm.html#sionna.ofdm.ResourceGridMapper) maps a sequence of constellation symbols onto a time-frequency resource grid.

If you simulate an AWGN or flat fading channel, there is really no need to define a complex resource grid for TDD as it simply means that you have two orthogonal dimensions for transmissions. You could simply simulate them independently without the need for any resource grid.

Hope that helps.

**Reply by neilb22** [reply]
Thank you very much, Jakob, for the informative and incredibly helpful answer. It is most appreciated.
